### ISSUE TITLE
Schema specification in root element only

### ISSUE THREAD

**Issue by SchweinchenFuntik** [pdvrieze/xmlutil#106]
Move all prefix declarations to the root element, or allow control over prefix setting (disable prefix insertion at the place of use).

**example model**:
```kotlin
@Serializable
@XmlSerialName("document", Namespaces.view, Namespaces.prefixView)
data class Document(
    val head: Head,
    val body: Body,
    val blocks: List<Block> = emptyList(),
    val nature: String? = null,
    val background: String? = null,
    val zoomable: Boolean? = null,
    @XmlOtherAttributes
    val other: Map<String, String> = emptyMap()
)
```

**example structure:**
```kotlin
    val document = Document(
        head = Head(
            variables = listOf(Variable("x", "internal"), Variable("y", "internal"),)
        ),
        body = Body(cdaTopic = "topic",
            elements = listOf(
                Div(listOf(If(condition = Expression("foo")))),
                Select(variable = "foo", elements = listOf(Select.Repeat(Repeat("test")))),
                Edit(Edit.Type.STRING, "foo"),
            )
        ),
        other = mapOf(Namespaces.xmlnsNaming.toCName() to Namespaces.naming)
    )
```

**output:**
```xml
<d:document xmlns:d="http://www.webui-ap.com/schemata/view/1.0" n:xmlns="http://www.webui-ap.com/schemata/naming/1.0">
  <d:head>
    <d:variable name="x" type="internal"/>
    <d:variable name="y" type="internal"/>
  </d:head>
  <d:body xmlns:n="http://www.webui-ap.com/schemata/naming/1.0" n:cda-topic="topic">
    <xhtml:div xmlns:xhtml="http://www.w3.org/1999/xhtml">
      <d:if condition="foo"/>
    </xhtml:div>
    <d:select variable="foo">
      <d:repeat from="test"/>
    </d:select>
    <d:edit variable="foo"/>
  </d:body>
</d:document>
```

**Desired view**
```xml
<d:document xmlns:d="http://www.webui-ap.com/schemata/view/1.0"
            n:xmlns="http://www.webui-ap.com/schemata/naming/1.0"
            xmlns:xhtml="http://www.w3.org/1999/xhtml">
    <d:head>
        <d:variable name="x" type="internal"/>
        <d:variable name="y" type="internal"/>
    </d:head>
    <d:body n:cda-topic="topic">
        <xhtml:div>
            <d:if condition="foo"/>
        </xhtml:div>
        <d:select variable="foo">
            <d:repeat from="test"/>
        </d:select>
        <d:edit variable="foo"/>
    </d:body>
</d:document>
```

**Comment by SchweinchenFuntik** [comment]
when specifying the default scheme, so that the prefix is not added at all

**Comment by pdvrieze** [comment]
When serializing it is possible to define the tagname (or prefix) for the outer element. If I'm not wrong I did implement a way to automatically put namespace tags on the outermost tag, but note that this is expensive as it effectively requires double serialization. If you know the prefixes you may want instead to have the outermost tag have a custom serializer that, for  the xml format checks that the encoder is an instance of XmlOutput and then retrieve the serializer from there (and add namespace tags explicitly - but only if that namespace isn't already present).

However the output you show here should not happen (repeating the already declared prefix - for the n prefix).

**Comment by pdvrieze** [comment]
The configuration attribute: `isCollectingNSAttributes` should do what you want (set it to true to have all the namespace declarations on the root).

**Comment by pdvrieze** [comment]
Btw. if you want to put something in the default namespace you can use the empty string for both. You can also set a namespace for unprefixed elements by specifying the prefix to be the empty string. On prefixes, note that the library will try to reuse prefixes before using the ones specified.

**Comment by SchweinchenFuntik** [comment]
`isCollectingNSAttributes = true`

```
Exception in thread "main" java.lang.StackOverflowError
	at kotlin.ranges.IntProgression.<init>(Progressions.kt:97)
	at kotlin.ranges.IntRange.<init>(Ranges.kt:44)
	at kotlin.ranges.RangesKt___RangesKt.until(_Ranges.kt:917)
	at nl.adaptivity.xmlutil.serialization.DefaultXmlSerializationPolicy.initialChildReorderMap(XmlSerializationPolicy.kt:464)
	at nl.adaptivity.xmlutil.serialization.structure.XmlCompositeDescriptor.<init>(XmlDescriptor.kt:604)
	at nl.adaptivity.xmlutil.serialization.structure.XmlDescriptor$Companion.from$xmlutil_serialization(XmlDescriptor.kt:271)
	at nl.adaptivity.xmlutil.serialization.structure.XmlDescriptor$Companion.from$xmlutil_serialization$default(XmlDescriptor.kt:209)
	at nl.adaptivity.xmlutil.serialization.structure.XmlCompositeDescriptor.createElementDescriptor(XmlDescriptor.kt:666)
	at nl.adaptivity.xmlutil.serialization.structure.XmlCompositeDescriptor.access$createElementDescriptor(XmlDescriptor.kt:583)
	at nl.adaptivity.xmlutil.serialization.structure.XmlCompositeDescriptor$children$2.invoke(XmlDescriptor.kt:613)
	at nl.adaptivity.xmlutil.serialization.structure.XmlCompositeDescriptor$children$2.invoke(XmlDescriptor.kt:607)
	at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
	at nl.adaptivity.xmlutil.serialization.structure.XmlCompositeDescriptor.getChildren(XmlDescriptor.kt:607)
	at nl.adaptivity.xmlutil.serialization.structure.XmlCompositeDescriptor.getElementDescriptor(XmlDescriptor.kt:670)
	at nl.adaptivity.xmlutil.serialization.XML.collectNamespaces$collect-2(XML.kt:295)
	at nl.adaptivity.xmlutil.serialization.XML.collectNamespaces$collect-2(XML.kt:299)
	at nl.adaptivity.xmlutil.serialization.XML.collectNamespaces$collect-2(XML.kt:299)
	at nl.adaptivity.xmlutil.serialization.XML.collectNamespaces$collect-2(XML.kt:299)
```

**Comment by SchweinchenFuntik** [comment]
Also
```xml
<d:document xmlns:d="http://www.webui-ap.com/schemata/view/1.0" n:xmlns="http://www.webui-ap.com/schemata/naming/1.0">
  <d:head>
    <d:variable name="x" type="internal"/>
    <d:variable name="y" type="internal"/>
  </d:head>
  <d:body xmlns:n="http://www.webui-ap.com/schemata/naming/1.0" n:cda-topic="topic">
    <xhtml:div xmlns:xhtml="http://www.w3.org/1999/xhtml">
      <d:if condition="foo"/>
    </xhtml:div>
    <d:select variable="foo"/>
    <d:edit variable="foo">
      <d:edit variable="foo2"/>
    </d:edit>
  </d:body>
</d:document>
```
**error**
```
Exception in thread "main" nl.adaptivity.xmlutil.XmlException: javax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,119]
Message: http://www.w3.org/TR/1999/REC-xml-names-19990114#AttributePrefixUnbound?d:document&n:xmlns&n
	at nl.adaptivity.xmlutil.StAXReader.next(StAXReader.kt:157)
	at nl.adaptivity.xmlutil.XmlReaderUtil__XmlReaderKt.skipPreamble(XmlReader.kt:381)
	at nl.adaptivity.xmlutil.XmlReaderUtil.skipPreamble(Unknown Source)
	at nl.adaptivity.xmlutil.serialization.XML.decodeFromReader(XML.kt:396)
	at nl.adaptivity.xmlutil.serialization.XML.decodeFromReader$default(XML.kt:383)
	at nl.adaptivity.xmlutil.serialization.XML.decodeFromString(XML.kt:349)
	at org.funtik.mscada.MainKt.main(Main.kt:66)
	at org.funtik.mscada.MainKt.main(Main.kt)
Caused by: javax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,119]
Message: http://www.w3.org/TR/1999/REC-xml-names-19990114#AttributePrefixUnbound?d:document&n:xmlns&n
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLStreamReaderImpl.next(XMLStreamReaderImpl.java:652)
	at nl.adaptivity.xmlutil.StAXReader.next(StAXReader.kt:155)
	... 7 more
```

**Comment by SchweinchenFuntik** [comment]
>  On prefixes, note that the library will try to reuse prefixes before using the ones specified.

cool, and even solves the problem, but I want `isCollectingNSAttributes`

**Comment by pdvrieze** [comment]
Could you provide a sample replication of the code causing the stack overflow? That should speed up replicating the bug.

**Comment by SchweinchenFuntik** [comment]
sorry for delay
```kotlin
package issue.bug

import kotlinx.serialization.*
import nl.adaptivity.xmlutil.serialization.*
import issue.bug.Repeat as RealRepeat
import issue.bug.Option as RealOption

sealed interface Element // Any

fun main() {
    val xml = XML {
        indent = 4
        autoPolymorphic = true
        encodeDefault = XmlSerializationPolicy.XmlEncodeDefault.NEVER
        isCollectingNSAttributes = true // BUG
    }
    issue106(xml)
}

fun issue106(xml: XML) {
    val document = Document(
        head = Head("ISSUE 106"),
        body = Body(
            elements = listOf(
                Select(variable = "foo", elements = listOf(Select.Repeat(RealRepeat("test")))),
            )
        )
    )
    val strXml = xml.encodeToString(document)
    println(strXml)
    val decodeDoc = xml.decodeFromString<Document>(strXml)
    println(decodeDoc)
}

@Serializable
@XmlSerialName("document", Namespaces.view, Namespaces.prefixView)
data class Document(
    val head: Head,
    val body: Body,
    val nature: String? = null, // DEFAULT XMLNS
    @XmlOtherAttributes
    val other: Map<String, String> = emptyMap()
)
@Serializable
@XmlSerialName("head", Namespaces.view, Namespaces.prefixView)
data class Head(@XmlElement(true) val title: String? = null)


@Serializable
@XmlSerialName("body",  Namespaces.view,  Namespaces.prefixView)
data class Body(
    val elements: List<Element> = emptyList(),
    @XmlOtherAttributes
    val other: Map<String, String> = emptyMap()
)

@Serializable
@XmlSerialName("select", Namespaces.view, Namespaces.prefixView)
data class Select(
    val variable: String = "",
    val elements: List<Node> = emptyList()
) : Element {

    @Serializable
    sealed interface Node // typing scheme
    @JvmInline @Serializable value class Repeat(val repeat: RealRepeat) : Node
    @JvmInline @Serializable value class Option(val repeat: RealOption) : Node
}

@Serializable
@XmlSerialName("repeat", Namespaces.view, Namespaces.prefixView)
data class Repeat(
    val text: String = "",
    val elements: List<Element> = emptyList()
) : Element

@Serializable
@XmlSerialName("option", Namespaces.view, Namespaces.prefixView)
data class Option(@Required val value: String, val text: String = "", val selected: String = "") : Element


object Namespaces {
    const val view = "http://custom"
    const val prefixView = "d"
}
```

**Comment by pdvrieze** [comment]
I've fixed this (and #107 that was triggered upon solving the base problem). The problem was an infinite recursion in the descriptor tree that wasn't considered possible (and thus not detected) - to collate namespace declarations the code walks the entire tree.

**Comment by pdvrieze** [comment]
Now released in [0.85.0](https://github.com/pdvrieze/xmlutil/releases/tag/v0.85.0).

**Comment by wbervoets** [comment]
Using 0.86.1 I'm always getting:

java.lang.IndexOutOfBoundsException: There are no children
	at nl.adaptivity.xmlutil.serialization.structure.XmlDescriptor.getElementDescriptor(XmlDescriptor.kt:151)
	at nl.adaptivity.xmlutil.serialization.XML.collectNamespaces$collect$5(XML.kt:293)
	at nl.adaptivity.xmlutil.serialization.XML.collectNamespaces$collect$5(XML.kt:303)
	at nl.adaptivity.xmlutil.serialization.XML.collectNamespaces(XML.kt:311)
	at nl.adaptivity.xmlutil.serialization.XML.encodeToWriter(XML.kt:216)
	at nl.adaptivity.xmlutil.serialization.XML.encodeToWriter(XML.kt:172)
	at nl.adaptivity.xmlutil.serialization.XML.encodeToString(XML.kt:117)
	at nl.adaptivity.xmlutil.serialization.XML.encodeToString(XML.kt:104)


when enabling isCollectingNSAttributes. Any idea's?

```
    public open fun getElementDescriptor(index: Int): XmlDescriptor {
        throw IndexOutOfBoundsException("There are no children")
    }
```
I think it fails due to my xml elements or attributes being an enum class. Is that possible?
